### ISSUE TITLE
Recipe for nvm?

### ISSUE THREAD

**Issue by binarykitchen** [Unitech/pm2#2171]
There is a problem and need advice/solution how to solve it.

You all know nvm, a great tool to select a node.js version. I have it installed on my server for the nodejs user. But pm2 during server restart runs on root user and currently does NOT tell nvm to load the default node.js version.

How do you do this? I think, in pm2-init.sh under init.d we need to add a pre-run script to call nvm first?

**Comment by fresheneesz** [comment]
There should be a way for you to change what user is used to start pm2 on server start. Then you can start pm2 with whatever user you're using nvm with.

**Comment by binarykitchen** [comment]
Nope, this didn't help.

So, I edited `pm2-init.sh` at the top for the nodejs user. First I source `nvm`, adjust some paths and then it worked like a charm after a server restart:

```                                                                                                                                        
source /home/nodejs/.nvm/nvm.sh

NODE_PATH=$NVM_BIN/node
PM2=$NVM_BIN/../lib/node_modules/pm2/bin/pm2

NAME=pm2
USER=nodejs
DEFAULT=/etc/default/$NAME

export PATH=/usr/bin:$PATH:$NODE_PATH
export PM2_HOME="/home/nodejs/.pm2"
```

I think this should be in the source code somehow. By detecting in the USER home whether nvm is installed and then source it!

**Comment by animedbz16** [comment]
Why not use the users .bashrc file to source the NVM path. This is how we
do this since nvm may not be installed in the same place for each user on
various OS environments.

**Comment by binarykitchen** [comment]
using .zshrc here. and running it during boot under root is probably a bad idea.

my point is: why don't you add an automatic nvm detection for the given USER?

**Comment by animedbz16** [comment]
I'm not sure I understand this. Your pm2-init.sh file which you included above is loading as root to start the pm2 service as the `nodejs` user. If pm2 is not starting properly through this service because your `nodejs` isn't aware of where pm2 is located then you should updates the `nodejs` user to source the NVM path from that user.

We don't have any issues with having PM2 automatically starting at boot and with running PM2 from a user which automatically sources NVM from within its own .bashrc file, which then loads the users default node environment which includes the pm2 executable installed globally under that version of node managed from NVM.

**Comment by binarykitchen** [comment]
Understand. One question for now. Why does it have to be .bashrc when that user only has .zshrc?

**Comment by animedbz16** [comment]
It doesn't have to be .bashrc. On our environment, bash is the default shell for users, so when this service starts, it attempts to execute pm as this user which will load up the user environment settings which will load up everything in .bashrc. If your environment uses the z shell (.zshrc) then those settings should likely go there. It depends on your system and settings.

For example, you can see my default shell is bash, so when root executes a command and logs into my user to do this, it will launch into the bash shell for this, so I need to use the .bashrc to define user defined environment settings to they will be loaded/available.

```
[user@media ~]$ cat /etc/passwd | grep user
user:x:1000:1000:user:/home/user:/bin/bash
```

**Comment by binarykitchen** [comment]
yep, got it and know that. still, doesnt work.

here are the steps to reproduce and please give it a try
1. create new user that uses zsh instead of bash
2. remove any nodejs binaries in /usr/bin (or remove via apt-get)
3. install latest nvm
4. install latest nodejs via nvm and set it to default
5. install latest pm2 globally
6. set up a small nodejs project using and running on pm2
7. save that project to pm2 (dump)
8. ensure that nodejs project is running fine
9. restart linux server
10. after few minutes that nodejs project is still NOT running (`pm2 ls` lists nothing)

**Comment by soyuka** [comment]
Do you have something like this in your `.zshrc`:
```
# Loads nvm
[ -s "${HOME}/.nvm/nvm.sh" ] && . "${HOME}/.nvm/nvm.sh"
```

Ensure that you've the startup script (or run `pm2 startup`).

**Comment by binarykitchen** [comment]
yes i have this at bottom of my .zshrc

```
export NVM_DIR="/home/nodejs/.nvm"
[ -s "$NVM_DIR/nvm.sh" ] && . "$NVM_DIR/nvm.sh"  # This loads nvm
```

of course, already did `pm2 startup` before, here evidence

```
$ sudo su -c "env PATH=$PATH:/home/nodejs/.nvm/versions/node/v6.1.0/bin pm2 startup ubuntu -u nodejs --hp /home/nodejs"                                                                                                               
[PM2] Spawning PM2 daemon
[PM2] PM2 Successfully daemonized
[PM2] Generating system init script in /etc/init.d/pm2-init.sh
[PM2] Making script booting at startup...
[PM2] -ubuntu- Using the command:
      su -c "chmod +x /etc/init.d/pm2-init.sh && update-rc.d pm2-init.sh defaults"

 System start/stop links for /etc/init.d/pm2-init.sh already exist.
[PM2] Done.
```

still, the above doesnt work and i have to manually tweak `pm2-init.sh` like mentioned above.

**Comment by evollu** [comment]
I'm getting the same issue, npm not found.
there is no issue if I directly ssh to that server, only when using pm2

I'm running ubuntu 16 (in azure) and have 
```
export NVM_DIR="/home/nodejs/.nvm"
[ -s "$NVM_DIR/nvm.sh" ] && . "$NVM_DIR/nvm.sh"  # This loads nvm
```
in ~/.profile

**Comment by vmarchaud** [comment]
Should be good with new startup system published with pm2 v2.2

**Comment by binarykitchen** [comment]
Still having those problems. What's really the recipe here when using nvm and zsh? Is there any documentation? In which zshrc file should I source nvm? Do not want to create any bash files at ll. cc @vmarchaud
